### PR TITLE
Limiting fitness evaluations

### PR DESCRIPTION
**Pull request by HenrikMettler**
This PR draft is a suggestion for a solution to #224 
It works by counting the calls of the objective function using a decorator (as suggested eg: https://stackoverflow.com/questions/21716940/is-there-a-way-to-track-the-number-of-times-a-function-is-called). I am not sure if this is a reasonable approach, mainly two things bother me: 

- the subtraction if fitness is not reevalutated (necessary since we don't want to count those)
- if a caching decorator is used for the inner objective, the counts of the objective function calls might not correspond to full function evaluations anymore

### CODE REVIEW

**Comment by jakobj** [comment]
thanks for working on this! didn't have time yet to check it in detail, but as i already mentioned we could think about introducing a general "objective" decorator, that would allow us to do such things like counting the number of objective evaluations in a simpler fashion. this would also remove the need to manually add the `if individual.fitness is None` statement to every objective. it would look something like:
```python
@cgp.objective
def objective(individual):
    ....
    return individual
```
what do you think @mschmidt87?

**Comment by mschmidt87** [comment]
Totally agree, a general objective wrapper sounds awesome. So, what you're saying is that within this PR, we would use @HenrikMettler 's code and
- add the `if individual.fitness is None` check to it
- move the decorator from the example to `utils.py`
, right?

**Comment by mschmidt87** [comment]
Little activity on this PR, @HenrikMettler are you working on this PR?

**Comment by HenrikMettler** [comment]
I moved the wrapper into `utils.py`.  Since I am not sure this is the way you were thinking @jakobj I didn't adapt all the examples. Also many tests in `test_hl_api.py` and `test_utils.py` fail under the current, because their `objectives` have no attribute `calls`

File: cgp/hl_api.py [hunk 1]
```diff
@@ -31,10 +32,12 @@ def evolve(
...
         The evolution algorithm. Needs to be a class instance with an
         `initialize_fitness_parents` and `step` method.
-    max_generations : int
-        Maximum number of generations.
     min_fitness : float
         Minimum fitness at which the evolution is stopped.
+    max_generations : int
+        Maximum number of generations. Defaults to inf
+    max_evaluations: int
+        Maximum number of function evaluations. Defaults to inf
```
Comment (jakobj) [inline cgp/hl_api.py:40]:
```suggestion
    max_generations : int
        Maximum number of generations. Defaults to positive infinity. Either this or `max_evaluations` need to be set to a finite value.
    max_evaluations: int
        Maximum number of function evaluations. Defaults to positive infinity.
```

File: cgp/hl_api.py [hunk 2]
```diff
@@ -49,6 +52,12 @@ def evolve(
     -------
     None
     """
+    if max_generations == np.inf and max_evaluations == np.inf:
```
Comment (jakobj) [inline cgp/hl_api.py:55]:
```suggestion
    if np.isinf(max_generations) and np.isinf(max_evaluations):
```

File: cgp/hl_api.py [hunk 3]
```diff
@@ -49,6 +52,12 @@ def evolve(
...
+        raise (
+            "Faulty configuration of evolve parameters: "
+            "Either max_generations or max_evaluation must be finite, "
+            "to avoid non terminal evolutionary process"
+        )
```
Comment (jakobj) [inline cgp/hl_api.py:60]:
```suggestion
        raise ValueError("Either `max_generations` or `max_evaluation` must be finite.")
```

File: cgp/hl_api.py [hunk 4]
```diff
@@ -69,11 +78,20 @@ def evolve(
             max_fitness = pop.champion.fitness
 
         if print_progress:
-            print(
-                f"\r[{pop.generation + 1}/{max_generations}] max fitness: {max_fitness}\033[K",
-                end="",
-                flush=True,
-            )
+            if not max_generations == np.inf:
```
Comment (jakobj) [inline cgp/hl_api.py:81]:
```suggestion
            if np.isfinite(max_generations):
```

File: cgp/hl_api.py [hunk 5]
```diff
@@ -69,11 +78,20 @@ def evolve(
...
+                print(
+                    f"\r[{pop.generation + 1}/{max_generations}] max fitness: {max_fitness}\033[K",
+                    end="",
+                    flush=True,
+                )
+            elif not max_evaluations == np.inf:
```
Comment (jakobj) [inline cgp/hl_api.py:87]:
```suggestion
            elif np.isfinite(max_evaluations):
```

File: cgp/hl_api.py [hunk 6]
```diff
@@ -69,11 +78,20 @@ def evolve(
...
+                print(
+                    f"\r[{objective.calls}/{max_evaluations}] max fitness: {max_fitness}\033[K",
```
Comment (jakobj) [inline cgp/hl_api.py:89]: doesn't this also need a `+1` like in the other if branch?

File: cgp/utils.py [hunk 7]
```diff
@@ -201,3 +201,26 @@ def primitives_from_class_names(primitives_str: Tuple[str, ...]) -> Tuple[Type[N
         primitives.append(primitives_dict[s])
 
     return tuple(primitives)
+
+
+def objective(f: Callable[[IndividualBase], IndividualBase]):
```
Comment (jakobj) [inline cgp/utils.py:206]:
```suggestion
def objective(f: Callable[[IndividualBase], IndividualBase]) -> Callable[[IndividualBase], IndividualBase]:
```

File: cgp/utils.py [hunk 8]
```diff
@@ -201,3 +201,26 @@ def primitives_from_class_names(primitives_str: Tuple[str, ...]) -> Tuple[Type[N
...
+    """ Wrapper for the objective function.
+
+    Wraps the objective function and counts the calls of the function.
```
Comment (jakobj) [inline cgp/utils.py:209]:
```suggestion
    """ Decorator for the objective function.
    
    Skips evaluations of individuals with fitness != None and counts number of objective calls.
```

File: cgp/utils.py [hunk 9]
```diff
@@ -201,3 +201,26 @@ def primitives_from_class_names(primitives_str: Tuple[str, ...]) -> Tuple[Type[N
...
+
+    Parameters
+    ----------
+    f: Callable
+        The objective function.
+    """
+
+    @functools.wraps(f)
+    def wrapped(individual: IndividualBase, *args, **kwargs):
```
Comment (jakobj) [inline cgp/utils.py:218]:
```suggestion
    def wrapped(individual: IndividualBase, *args, **kwargs) -> IndividualBase:
```

File: cgp/utils.py [hunk 10]
```diff
@@ -201,3 +201,26 @@ def primitives_from_class_names(primitives_str: Tuple[str, ...]) -> Tuple[Type[N
...
+        if individual.fitness is None:
+            wrapped.calls += 1
+            return f(individual, *args, **kwargs)
+        else:
+            return individual
+
+    wrapped.calls = 0
```
Comment (jakobj) [inline cgp/utils.py:225]:
```suggestion
    wrapped.calls: int = 0
```

File: cgp/utils.py [hunk 11]
```diff
@@ -201,3 +201,26 @@ def primitives_from_class_names(primitives_str: Tuple[str, ...]) -> Tuple[Type[N
...
+
+def objective(f: Callable[[IndividualBase], IndividualBase]):
+    """ Wrapper for the objective function.
+
+    Wraps the objective function and counts the calls of the function.
+
+    Parameters
+    ----------
+    f: Callable
+        The objective function.
```
Comment (jakobj) [inline cgp/utils.py:214]:
```suggestion
    f : Callable
        The objective function.
```

File: examples/example_minimal.py [hunk 12]
```diff
@@ -41,9 +41,8 @@ def f_target(x):
 # on a set of random points.
 
 
+@cgp.utils.objective
```
Comment (jakobj) [inline examples/example_minimal.py:44]: as we'll be using this in every script, maybe pull it up into the cgp namespace by importing it in `cgp/__init__.py`? that would allows us to write `@cgp.objective` which seems nicer.

**Review by jakobj — CHANGES_REQUESTED** [review]
thanks for the changes :ok_hand: i've left some inline comments that should be addressed.

File: cgp/hl_api.py (hunk 6 above)
Comment (HenrikMettler) [reply cgp/hl_api.py:89]: I don't think so, since the `objective.calls` has no offset to the number of evaluations. What I just realized could happen though, is that there are too many `evaluations`. Since `ea.step` (https://github.com/HenrikMettler/hal-cgp/blob/212bdea9dd53207ba0f0ec097d2596d0bd2f9767/cgp/hl_api.py#L71) can call the `objective` function more than once (for multiple offspring with different phenotype). I don't know if there is an easy fix to this without moving the while loop for `objective.calls` into the `ea.step`

**Comment by HenrikMettler** [comment]
Should I introduce the objective decorator for all examples and tests within this PR, or in a separate PR?

**Comment by jakobj** [comment]
> Should I introduce the objective decorator for all examples and tests within this PR, or in a separate PR?

let's first figure out a suitable implementation that works with more than one process. we might end up with a different solution than the decorator after all.

**Comment by HenrikMettler** [comment]
As suggested by @jakobj the counter for objective function calls is now moved to the evolutionary algorithm. Removing it from the `objective` decorator enables using it with multiple processes running in parallel. I left everything for the `objective` decorator within this PR, even though it doesn't really belong to this topic anymore

File: cgp/ea/mu_plus_lambda.py [hunk 13]
```diff
@@ -52,6 +52,8 @@ def __init__(
         self.local_search = local_search
         self.k_local_search = k_local_search
 
+        self.objective_calls = 0
```
Comment (jakobj) [inline cgp/ea/mu_plus_lambda.py:55]:
```suggestion
        self.n_objective_calls = 0
```
to indicate that this is a number. we use it in a couple of places, so why not here as well as it can increase readability.

File: cgp/ea/mu_plus_lambda.py [hunk 14]
```diff
@@ -102,6 +104,9 @@ def step(
         # population instead of the other way around
         combined = offsprings + pop.parents
 
+        # before computing the fitness update the count of objective function calls
+        self.update_objective_counts(combined)
```
Comment (jakobj) [inline cgp/ea/mu_plus_lambda.py:108]:
```suggestion
        self.update_n_objective_calls(combined)
```

File: cgp/ea/mu_plus_lambda.py [hunk 15]
```diff
@@ -102,6 +104,9 @@ def step(
         # population instead of the other way around
         combined = offsprings + pop.parents
 
+        # before computing the fitness update the count of objective function calls
```
Comment (jakobj) [inline cgp/ea/mu_plus_lambda.py:107]:
```suggestion
```

File: cgp/ea/mu_plus_lambda.py [hunk 16]
```diff
@@ -182,3 +188,8 @@ def _create_new_parent_population(
 
         """
         return combined[:n_parents]
+
+    def update_objective_counts(self, combined: List[IndividualBase]) -> None:
```
Comment (jakobj) [inline cgp/ea/mu_plus_lambda.py:192]:
```suggestion
    def update_objective_counts(self, combined: List[IndividualBase]) -> None:
        """Increase n_objective_calls by the number of individuals with fitness=None, i.e., for which the objective function will be evaluated.
        """ 
```

File: cgp/hl_api.py [hunk 17]
```diff
@@ -31,10 +32,15 @@ def evolve(
...
         The evolution algorithm. Needs to be a class instance with an
         `initialize_fitness_parents` and `step` method.
-    max_generations : int
-        Maximum number of generations.
     min_fitness : float
         Minimum fitness at which the evolution is stopped.
+    max_generations : int
+        Maximum number of generations.
+        Defaults to positive infinity.
+        Either this or `max_evaluations` need to be set to a finite value.
```
Comment (jakobj) [inline cgp/hl_api.py:40]:
```suggestion
        Either this or `max_objective_calls` needs to be set to a finite value.
```

File: cgp/utils.py [hunk 18]
```diff
@@ -201,3 +201,25 @@ def primitives_from_class_names(primitives_str: Tuple[str, ...]) -> Tuple[Type[N
         primitives.append(primitives_dict[s])
 
     return tuple(primitives)
+
+
+def objective(
```
Comment (jakobj) [inline cgp/utils.py:206]: i would remove this from this PR, we don't use it for keeping track of the number of objective function evaluations, do we?

File: examples/example_minimal.py [hunk 19]
```diff
@@ -82,7 +80,8 @@ def objective(individual):
 
 ea_params = {"n_offsprings": 4, "tournament_size": 2, "n_processes": 2}
 
-evolve_params = {"max_generations": int(args["--max-generations"]), "min_fitness": 0.0}
+evolve_params = {"max_generations": np.inf, "min_fitness": 0.0, "max_objective_calls": 10000}
```
Comment (jakobj) [inline examples/example_minimal.py:83]: is this change intentional or remaining from testing? if intentional, what's the motivation?

**Review by jakobj — CHANGES_REQUESTED** [review]
thanks for (re)implementing this so quickly! :rocket: 

i've added a few inline comments. also please add a unit test.

File: examples/example_minimal.py (hunk 19 above)
Comment (HenrikMettler) [reply examples/example_minimal.py:83]: It is still from testing. Should the new functionality be in a separate example (or just available and covered by tests)?

Comment (jakobj) [reply examples/example_minimal.py]: i'd suggest to just have it available, documented and covered by tests. no need for an additional example.

File: test/test_ea_mu_plus_lambda.py [hunk 20]
```diff
@@ -221,3 +221,46 @@ def test_create_new_parent_population(population_params, genome_params, ea_param
...
+def test_update_n_objective_calls(population_params, genome_params, ea_params):
+    def objective(individual):
+        individual.fitness = float(individual.idx)
+        return individual
+
+    # create a population and an instance of the evolutionary algorithm and
+    # check that number of objective calls is initially zero
+    pop = cgp.Population(**population_params, genome_params=genome_params)
+    ea = cgp.ea.MuPlusLambda(**ea_params)
+    assert ea.n_objective_calls == 0
```
Comment (jakobj) [inline test/test_ea_mu_plus_lambda.py:235]:
```suggestion
    # check that number of objective calls is initially zero
    n_objective_calls_expected = 0
    pop = cgp.Population(**population_params, genome_params=genome_params)
    ea = cgp.ea.MuPlusLambda(**ea_params)
    assert ea.n_objective_calls == n_objective_calls_expected
```
i would introduce this `_expected` variable and set it always according to what you expect such that you can use it in the `assert`s. i think this would make the code clearer and also the comments obsolete.

File: test/test_ea_mu_plus_lambda.py [hunk 21]
```diff
@@ -221,3 +221,46 @@ def test_create_new_parent_population(population_params, genome_params, ea_param
...
+    n_objective_calls = ea.n_objective_calls
+    n_generations = 100
+    for _ in range(n_generations):
+        offsprings = ea._create_new_offspring_generation(pop)
+        combined = offsprings + pop.parents
+        n_objective_calls += sum([1 for ind in combined if ind.fitness is None])
+        combined = ea._compute_fitness(combined, objective)
+        assert n_objective_calls == ea.n_objective_calls
+
+    # create a population with mutation rate one and check that the number
```
Comment (jakobj) [inline test/test_ea_mu_plus_lambda.py:253]: nice idea for a second test! but i'd suggest to move it to a separate `test_` function with an appropriate name.

File: test/test_hl_api.py [hunk 22]
```diff
@@ -135,6 +135,22 @@ def f1(x):
     assert abs(pop.champion.fitness) == pytest.approx(0.0)
 
 
+def test_finite_evolve(population_params, genome_params, ea_params):
```
Comment (jakobj) [inline test/test_hl_api.py:138]:
```suggestion
def test_finite_max_generations_or_max_objective_calls(population_params, genome_params, ea_params):
```

File: test/test_hl_api.py [hunk 23]
```diff
@@ -135,6 +135,22 @@ def f1(x):
...
+    def objective(individual):
+        individual.fitness = float(individual.idx)
+        return individual
+
+    pop = cgp.Population(**population_params, genome_params=genome_params)
+    ea = cgp.ea.MuPlusLambda(**ea_params)
+    evolve_params = {
+        "max_generations": np.inf,
+        "min_fitness": -1e-12,
```
Comment (jakobj) [inline test/test_hl_api.py:147]:
```suggestion
        "min_fitness": 0.0,
```
the loop exit criterion takes floating point imprecision into account, no need for such funky fitness values ;)

**Review by jakobj — CHANGES_REQUESTED** [review]
thanks for the update! :ok_hand: 

some more inline comments, then it's ready to merge!

File: test/test_ea_mu_plus_lambda.py [hunk 24]
```diff
@@ -228,39 +228,37 @@ def objective(individual):
...
-    # to the number of parents
     ea.initialize_fitness_parents(pop, objective)
-    assert ea.n_objective_calls == population_params["n_parents"]
+    n_objective_calls_expected = population_params["n_parents"]
+    assert ea.n_objective_calls == n_objective_calls_expected
 
-    # count the number of objective calls by adding the number of offspring with fitness = None
-    # at each generation
-    n_objective_calls = ea.n_objective_calls
+    n_objective_calls_expected = ea.n_objective_calls
```
Comment (jakobj) [inline test/test_ea_mu_plus_lambda.py:240]:
```suggestion
```
this is not necessary as you've just checked that they are equal, right?

**Review by jakobj — APPROVED** [review]
excellent, thanks a lot! :+1:

as always, please reorganize the history before merging.
